corehq/apps/accounting/invoicing.py

```diff
@@ -353,6 +273,150 @@ def create_wire_credits_invoice(self, items, amount):
         )
 
 
+class CustomerAccountInvoiceFactory(object):
+    """
+        This generates an invoice for a Customer Billing Account.
+    """
+    def __init__(self, date_start, date_end, account, recipients=None):
+        """
+        The Invoice generated will always be for the month preceding the
+        invoicing_date.
+        For example, if today is July 5, 2014 then the invoice will be from
+        June 1, 2014 to June 30, 2014.
+        """
+        self.date_start = date_start
+        self.date_end = date_end
+        self.account = account
+        self.recipients = recipients
+        self.customer_invoice = None
+        self.subscriptions = {}
+
+    def create_invoice(self):
+        for subscription in self.account.subscription_set.all():
+            if should_create_invoice(subscription, subscription.subscriber.domain, self.date_start, self.date_end):
+                self.subscriptions[subscription.plan_version] = subscription
+        if not self.subscriptions:
+            return
+        self._generate_customer_invoice()
+        self._email_invoice()
+
+    def _generate_customer_invoice(self):
+        plan_version, subscription = self.subscriptions.popitem()
+        invoice = generate_invoice(subscription, self.date_start, self.date_end)
+        for plan in self.subscriptions:
+            generate_line_items(invoice, self.subscriptions[plan])
+            invoice.calculate_credit_adjustments()
+            invoice.update_balance()
+            invoice.save()
+        update_invoice_due_date(invoice, subscription, self.date_end)
+        self.customer_invoice = invoice
+
+    def _email_invoice(self):
+        record = BillingRecord.generate_record(self.customer_invoice)
+        try:
+            if self.recipients:
+                for email in self.recipients:
+                    record.send_email(contact_email=email)
+            elif self.account.dimagi_contact:
+                record.send_email(contact_email=self.account.dimagi_contact,
+                                  cc_emails=[settings.ACCOUNTS_EMAIL])
+            else:
+                record.send_email(contact_email=settings.ACCOUNTS_EMAIL)
+        except InvoiceEmailThrottledError as e:
+            log_accounting_error(str(e))
+
+
+def should_create_invoice(subscription, domain, invoice_start, invoice_end):
+    if subscription.is_trial:
+        log_accounting_info("Skipping invoicing for Subscription %s because it's a trial." % subscription.pk)
+        return False
+    if subscription.skip_invoicing_if_no_feature_charges and not \
+            subscription.plan_version.feature_charges_exist_for_domain(domain):
+        log_accounting_info(
+            "Skipping invoicing for Subscription %s because there are no feature charges."
+            % subscription.pk
+        )
+        return False
+    if subscription.date_start >= invoice_end:
```

Comment:
@gbova Why is this ```>=``` instead of ```>```?  I think this resulted in subscriptions that start on the last day of the month not getting invoiced.

Reply:
Ah I see that now. I think you're right, this should be `>`.